source/fundamentals/csfle.txt

```diff
@@ -0,0 +1,224 @@
+     const client = new MongoClient("mongodb://localhost:27017", {
+       useNewUrlParser: true,
+       useUnifiedTopology: true,
+       autoEncryption: {
+         keyVaultNamespace: "encryption.__keyVault",
+         kmsProviders: {
+           local: {
+             key: fs.readFileSync("./master-key.txt"),
+           },
+         },
+         schemaMap,
+       },
+     });
+     return await client.connect();
+   }
+   
+   function createJsonSchemaMap(dataKey) {
+     return {
+       "users.ssns": {
+         bsonType: "object",
+         encryptMetadata: {
+           keyId: [new Binary(Buffer.from(dataKey, "base64"), 4)],
+         },
+         properties: {
+           ssn: {
+             encrypt: {
+               bsonType: "int",
+               algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic",
+             },
+           },
+         },
+       },
+     };
+   }
+   
+   async function makeDataKey(client) {
+     const encryption = new ClientEncryption(client, {
+       keyVaultNamespace: "encryption.__keyVault",
+       kmsProviders: {
+         local: {
+           key: fs.readFileSync("./master-key.txt"),
+         },
+       },
+     });
+     
+     let dataKey = await encryption.createDataKey("local", {
+       masterKey: null,
+     });
+     
+     return dataKey.toString("base64");
+   }
+   
+   async function run(regularClient, csfleClient) {
+     try {
+   
+       regularClient = await getRegularClient();
+   
+       let dataKey = await makeDataKey(regularClient);
+       console.log(
+         "New dataKey created for this run:\n",
+         dataKey
+       );
+   
+       const schemaMap = createJsonSchemaMap(dataKey);
+       
+       csfleClient = await getCsfleEnabledClient(schemaMap);
+   
+       const regularClientSsnsColl = regularClient
+         .db("users")
+         .collection("ssns");
+       const csfleClientSsnsColl = csfleClient
+         .db("users")
+         .collection("ssns");
+   
+       const exampleDocument = {
+         name: "Jon Doe",
+         ssn: 241014209,
+       };
+   
+       await csfleClientSsnsColl.updateOne(
+         { ssn: exampleDocument["ssn"] },
+         { $set: exampleDocument },
+         { upsert: true }
+       );
+   
+       const csfleFindResult = await csfleClientSsnsColl.findOne({
+         ssn: exampleDocument["ssn"],
+       });
+       console.log(
+         "Document retrieved with csfle enabled client:\n",
+         csfleFindResult
+       );
+   
+       const regularFindResult = await regularClientSsnsColl.findOne({
+         name: "Jon Doe",
+       });
+       console.log(
+         "Document retrieved with regular client:\n", 
+         regularFindResult
+       );
+   
+     } finally {
+       if (regularClient) await regularClient.close();
+       if (csfleClient) await csfleClient.close();
+     }
+   }
+   
+   let regularClient = null;
+   let csfleClient = null;
+   run(regularClient, csfleClient).catch(console.dir);
```

Comment:
```suggestion
run().catch(error => {
   console.dir(error);
   process.exit(1);
});
```
These are in the top level of the file right? the indentation looked off. Also the variables here are just null when passed in, I'm not sure I see the use of keeping them?

Reply:
They are, but the whole example is in a code block - I think the indentation is correct.

I think you're right about the null variables - I had thought I needed them in the global scope for some reason. Took your suggestion!